pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/RangeIndexCreator.java

```diff
@@ -0,0 +1,471 @@
+        int sizeInBytes = bitmap.serializedSizeInBytes();
+        bitmapOffset += sizeInBytes;
+
+        // Check for int overflow
+        Preconditions.checkState(bitmapOffset > 0, "Inverted index file: %s exceeds 2GB limit", _invertedIndexFile);
+
+        header.writeLong(bitmapOffset);
+        bytesWritten += Long.BYTES;
+
+        byte[] bytes = new byte[sizeInBytes];
+        bitmap.serialize(ByteBuffer.wrap(bytes));
+        dataOutputStream.write(bytes);
+        bytesWritten += bytes.length;
+      }
+    } catch (IOException e) {
+      FileUtils.deleteQuietly(_invertedIndexFile);
+      throw e;
+    }
+    Preconditions.checkState(bytesWritten == _invertedIndexFile.length(),
+        "Length of inverted index file: " + _invertedIndexFile.length()
+            + " does not match the number of bytes written: " + bytesWritten);
+  }
+
+  @Override
+  public void close()
+      throws IOException {
+    org.apache.pinot.common.utils.FileUtils
+        .close(new DataBufferAndFile(_forwardIndexValueBuffer, _forwardIndexValueBufferFile),
+            new DataBufferAndFile(_docIdValueBuffer, _docIdBufferFileForSorting));
+  }
+
+  private class DataBufferAndFile implements Closeable {
+    private final PinotDataBuffer _dataBuffer;
+    private final File _file;
+
+    DataBufferAndFile(final PinotDataBuffer buffer, final File file) {
+      _dataBuffer = buffer;
+      _file = file;
+    }
+
+    @Override
+    public void close()
+        throws IOException {
+      destroyBuffer(_dataBuffer, _file);
+    }
+  }
+
+  void dump() {
+    System.out.print("[ ");
+    for (int i = 0; i < _numValues; i++) {
+      System.out.print(_valueBuffer.get(_docIdBuffer.get(i).intValue()) + ", ");
+    }
+    System.out.println("]");
+  }
+
+  private PinotDataBuffer createTempBuffer(long size, File mmapFile)
+      throws IOException {
+    if (_useMMapBuffer) {
+      return PinotDataBuffer.mapFile(mmapFile, false, 0, size, PinotDataBuffer.NATIVE_ORDER,
+          "OffHeapBitmapInvertedIndexCreator: temp buffer");
+    } else {
+      return PinotDataBuffer.allocateDirect(size, PinotDataBuffer.NATIVE_ORDER,
+          "OffHeapBitmapInvertedIndexCreator: temp buffer for " + mmapFile.getName());
+    }
+  }
+
+  private void destroyBuffer(PinotDataBuffer buffer, File mmapFile)
+      throws IOException {
+    if (buffer != null) {
+      buffer.close();
+      if (mmapFile.exists()) {
+        FileUtils.forceDelete(mmapFile);
+      }
+    }
+  }
+
+  interface NumberValueBuffer {
+
+    void put(int position, Number value);
+
+    Number get(int position);
+
+    int compare(Number val1, Number val2);
+  }
+
+  class IntValueBuffer implements NumberValueBuffer {
+
+    private PinotDataBuffer _buffer;
+
+    IntValueBuffer(PinotDataBuffer buffer) {
+
+      _buffer = buffer;
+    }
+
+    @Override
+    public void put(int position, Number value) {
+      _buffer.putInt(position << 2, value.intValue());
+    }
```

Comment:
Using constants for << 2, << 3 would be better for readability. Something like INT_SIZE_BYTES, DOUBLE_SIZE_BYTES etc